internal/rules/resources_test.go

```diff
@@ -37,50 +37,64 @@ func TestRules(t *testing.T) {
 		NoError(t, err, "expect no error")
 		Assert(t, len(grants) == 2, "expected 2 grants")
-		Assert(t, grants[0].RoleID == 2, "expected RoleID == 2, got %v", grants[0].RoleID)
-		Assert(t, grants[0].Resource == "channel:2", "expected Resource == channel:2, got %s", grants[0].Resource)
-		Assert(t, grants[0].Operation == "delete", "expected Operation == delete, got %s", grants[0].Operation)
-		Assert(t, grants[0].Value == rules.Allow, "expected Value == Allow, got %s", grants[0].Value)
-	}
 
-	// list all by role
-	{
-		grants, err := resources.List(2)
-		NoError(t, err, "expected no error")
-		Assert(t, len(grants) == 2, "expected grants == 2, got %v", len(grants))
+		for _, grant := range grants {
+			Assert(t, grant.RoleID == 2, "expected RoleID == 2, got %v", grant.RoleID)
+			Assert(t, grant.Resource == "channel:2", "expected Resource == channel:2, got %s", grant.Resource)
+			// Assert(t, grant.Operation == "delete", "expected Operation == delete, got %s", grant.Operation)
+			Assert(t, grant.Value == rules.Allow, "expected Value == Allow, got %s", grant.Value)
+		}
 	}
 
 	// deny channel:1 group:1 (explicit deny, multi=deny)
 	{
-		resources.GrantByResource(1, "channel:1", []string{"edit"}, rules.Deny)
-		Expect(rules.Deny, resources.IsAllowed("channel:1", "edit"), "expected error, got nil")
-		Expect(rules.Allow, resources.IsAllowed("channel:2", "edit"), "channel:2 edit, expected no error")
-		Expect(rules.Deny, resources.IsAllowed("channel:*", "edit"), "expected error, got nil")
+		list := []rules.Rule{
+			rules.Rule{Resource: "channel:1", Operation: "update", Value: rules.Deny},
+		}
+		resources.Grant(1, list)
+		Expect(rules.Deny, resources.IsAllowed("channel:1", "update"), "expected error, got nil")
+		Expect(rules.Allow, resources.IsAllowed("channel:2", "update"), "channel:2 update, expected no error")
+		Expect(rules.Deny, resources.IsAllowed("channel:*", "update"), "expected error, got nil")
 	}
 
 	// reset (unset=deny)
 	{
-		resources.GrantByResource(2, "channel:2", []string{"edit", "delete"}, rules.Inherit)
-		resources.GrantByResource(1, "channel:1", []string{"edit", "delete"}, rules.Inherit)
-		Expect(rules.Inherit, resources.IsAllowed("channel:1", "edit"), "expected error, got nil")
-		Expect(rules.Inherit, resources.IsAllowed("channel:*", "edit"), "expected error, got nil")
+		list1 := []rules.Rule{
+			rules.Rule{Resource: "channel:1", Operation: "update", Value: rules.Inherit},
+			rules.Rule{Resource: "channel:1", Operation: "delete", Value: rules.Inherit},
+		}
+		resources.Grant(1, list1)
+
+		list2 := []rules.Rule{
+			rules.Rule{Resource: "channel:2", Operation: "update", Value: rules.Inherit},
+			rules.Rule{Resource: "channel:2", Operation: "delete", Value: rules.Inherit},
+		}
+		resources.Grant(2, list2)
+
+		Expect(rules.Inherit, resources.IsAllowed("channel:1", "update"), "expected error, got nil")
+		Expect(rules.Inherit, resources.IsAllowed("channel:*", "update"), "expected error, got nil")
```

Comment:
I would go so far to disallow (panic) IsAllowed checking against a _wildcarded_ resource. I know it is technically possible but in practice it will be (should not be) ever done, true?

Reply:
Okey will fix this on #48 together with validation and granting logic.